### ISSUE TITLE
Initiator host's nvme-cli hangs after attachment of 3 or 4 nvmf subsystems

### ISSUE THREAD

**Issue by Roberciiik** [spdk/spdk#677]
Hi,

I have tried to attach 10 NVMe-oF volumes to initiator host. It is done by creating nvmf_subsystem and then adding allowed host NQN to it. After each creation of nvmf_subsystem on storage service host I used nvme-cli to discover and connect previously created nvme-of target device. However after 3 or 4 (multiple run) attached devices on initiator host the "nvme connect" command hangs and "nvme list" command returns exit code 7 or 12 or hangs also.

I run this setup configuration about 4 times and the issue occurred in every run.

Thanks,
Robert

## Steps to Reproduce
1. Create 10 volumes using `construct_lvol_bdev` in SPDK.
2. Call SPDK `nvmf_subsystem_create` command with generated NQN (also `nvmf_subsystem_add_listener` and `nvmf_subsystem_add_ns`)
3. Call SPDK `nvmf_subsystem_add_host` command with initiator host's NQN.
4. Call nvme-cli `nvme discover` command from initiator host.
5. Call nvme-cli `nvme connect` command from initiator host.
6. Verify if new drive is attached by calling nvme-cli `nvme list` command from initiator host.
7. Repeat steps 2-6 nine times

## Current Behavior
Only 3 NVMf subsystems (NVMe drives) are visible on initiator host OS.
`nvme list` return exit code 7 or 12 or it hangs.
`nvme connect` hangs.

## Expected Behavior
All NVMf target subsystems are visible under initiator host OS.

## Possible Solution
Unknown.

## Context (Environment including OS version, SPDK version, etc.)
- nvme cli version: 1.5
- SPDK version 19.01
- Target & initiator hosts OS: Ubuntu 16.04

## Logs from dmesg of initiator host

```
[   53.557420] mlx5_core 0000:5e:00.0 enp94s0f0: Link up
[   54.228422] mlx5_core 0000:5e:00.1 enp94s0f1: Link up
[   54.653146] cgroup: new mount options do not match the existing superblock, will be ignored
[   54.906755] nvme nvme1: queue_size 128 > ctrl maxcmd 1, clamping down
[   54.906760] nvme nvme1: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[   55.976147] nvme nvme1: rdma_resolve_addr wait failed (-110).
[   56.013872] nvme nvme1: creating 8 I/O queues.
[   56.232982] nvme nvme1: new ctrl: NQN "nqn.2014-08.org.nvmexpress:uuid:749cb305-8b4d-45ad-ba3f-3be2d83bdce4", addr 10.2.0.46:4420
[   58.329253] ipmi device interface
[   61.168327] nvme nvme2: queue_size 128 > ctrl maxcmd 1, clamping down
[   61.168332] nvme nvme2: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[   66.131917] nvme nvme2: queue_size 128 > ctrl maxcmd 1, clamping down
[   66.131923] nvme nvme2: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450

[...] Same discovery logs... [...]

[  263.615679] nvme nvme2: queue_size 128 > ctrl maxcmd 1, clamping down
[  263.615684] nvme nvme2: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[  263.675295] nvme nvme2: Connect Invalid Data Parameter, subsysnqn "c8623707-2d1d-4bbb-85cd-e45723cc6bfb"
[  263.725677] nvme nvme2: creating 8 I/O queues.
[  263.945399] nvme nvme2: new ctrl: NQN "nqn.2014-08.org.nvmexpress:uuid:c8623707-2d1d-4bbb-85cd-e45723cc6bfb", addr 10.2.0.46:4420
[  268.996649] nvme nvme3: queue_size 128 > ctrl maxcmd 1, clamping down
[  268.996654] nvme nvme3: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[  274.064252] nvme nvme3: queue_size 128 > ctrl maxcmd 1, clamping down
[  274.064257] nvme nvme3: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450

[...] Same discovery logs... [...]

[  481.781950] nvme nvme3: queue_size 128 > ctrl maxcmd 1, clamping down
[  481.781955] nvme nvme3: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[  481.837496] nvme nvme3: Connect Invalid Data Parameter, subsysnqn "65736149-e3c0-4aa1-92e8-db62c80632e8"
[  481.887739] nvme nvme3: creating 8 I/O queues.
[  482.107159] nvme nvme3: new ctrl: NQN "nqn.2014-08.org.nvmexpress:uuid:65736149-e3c0-4aa1-92e8-db62c80632e8", addr 10.2.0.46:4420
[  487.159000] nvme nvme4: queue_size 128 > ctrl maxcmd 1, clamping down
[  487.159006] nvme nvme4: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450

[...] Same discovery logs... [...]

[  725.135774] nvme nvme4: queue_size 128 > ctrl maxcmd 1, clamping down
[  725.135778] nvme nvme4: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[  730.198359] nvme nvme4: queue_size 128 > ctrl maxcmd 1, clamping down
[  730.198364] nvme nvme4: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 10.2.0.100:4450
[  769.320167] nvme nvme1: reconnecting in 10 seconds
[  779.678315] nvme nvme1: Successfully reconnected
[  980.674024] nvme nvme2: reconnecting in 10 seconds
[  990.873010] nvme nvme2: Successfully reconnected
[ 1201.956629] nvme nvme3: reconnecting in 10 seconds
[ 1212.316004] nvme nvme3: Successfully reconnected
[ 1252.347919] nvme nvme2: reconnecting in 10 seconds
[ 1252.517010] nvme nvme3: reconnecting in 10 seconds
[ 1252.675253] nvme nvme1: reconnecting in 10 seconds
[ 1262.408129] nvme nvme2: Connect Invalid Data Parameter, subsysnqn "nqn.2014-08.org.nvmexpress:uuid:c8623707-2d1d-4bbb-85cd-e45723cc6bfb"
[ 1262.408133] nvme nvme2: Failed reconnect attempt, requeueing...
[ 1262.658106] nvme nvme3: Connect Invalid Data Parameter, subsysnqn "nqn.2014-08.org.nvmexpress:uuid:65736149-e3c0-4aa1-92e8-db62c80632e8"
[ 1262.658109] nvme nvme3: Failed reconnect attempt, requeueing...
[ 1262.917774] nvme nvme1: Connect Invalid Data Parameter, subsysnqn "nqn.2014-08.org.nvmexpress:uuid:749cb305-8b4d-45ad-ba3f-3be2d83bdce4"
[ 1262.917778] nvme nvme1: Failed reconnect attempt, requeueing...
[ 1272.628195] nvme nvme2: Connect Invalid Data Parameter, subsysnqn "nqn.2014-08.org.nvmexpress:uuid:c8623707-2d1d-4bbb-85cd-e45723cc6bfb"
[ 1272.628199] nvme nvme2: Failed reconnect attempt, requeueing...
[ 1272.878493] nvme nvme3: Connect Invalid Data Parameter, subsysnqn "nqn.2014-08.org.nvmexpress:uuid:65736149-e3c0-4aa1-92e8-db62c80632e8"
[ 1272.878497] nvme nvme3: Failed reconnect attempt, requeueing...
[ 1273.137883] nvme nvme1: Connect Invalid Data Parameter, subsysnqn "nqn.2014-08.org.nvmexpress:uuid:749cb305-8b4d-45ad-ba3f-3be2d83bdce4"

[...] Same failure logs... [...]

```

EDIT:
## Logs from SPDK -L all (from last 1000 lines)
```
[...] Repetitively [...]
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e1c0 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e1c0 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 1
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 5
request.c: 121:nvmf_trace_command: *DEBUG*: Admin cmd: opc 0x18 fuse 0 cid 0 nsid 0 cdw10 0x00000000
ctrlr.c:1764:spdk_nvmf_ctrlr_keep_alive: *DEBUG*: Keep Alive
request.c:  92:spdk_nvmf_request_complete: *DEBUG*: cpl: cid=0 cdw0=0x00000000 rsvd1=0 status=0x0000
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 7
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 9
rdma.c: 978:request_transfer_out: *DEBUG*: RDMA RECV POSTED. Recv: 0x26626f8 Connection: 0x2939d30
rdma.c:1001:request_transfer_out: *DEBUG*: RDMA SEND POSTED. Request: 0x28e02f0 Connection: 0x2939d30
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 11
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 11
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 12
rdma.c:1508:spdk_nvmf_rdma_request_process: *DEBUG*: Request 0x28e02f0 entering state 0
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e1c0 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e1c0 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e1c0 returned -1
reactor.c: 187:get_rusage: *INFO*: Reactor 0: 0 voluntary context switches and 3 involuntary context switches in the last second.
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x23613f0 returned -1
thread.c: 387:spdk_thread_poll: *DEBUG*: Timed poller 0x235e170 returned -1
[...]
```

## Logs from SPDK only ERRORs 
```
[...] Repetitively [...]
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'nqn.2014-08.org.nvmexpress:uuid:676fb590-f928-496c-8e50-9378179ccbd8'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'nqn.2014-08.org.nvmexpress:uuid:84e449a0-9328-48cd-9116-8fe63b268f14'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'nqn.2014-08.org.nvmexpress:uuid:676fb590-f928-496c-8e50-9378179ccbd8'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'nqn.2014-08.org.nvmexpress:uuid:84e449a0-9328-48cd-9116-8fe63b268f14'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'nqn.2014-08.org.nvmexpress:uuid:676fb590-f928-496c-8e50-9378179ccbd8'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'nqn.2014-08.org.nvmexpress:uuid:84e449a0-9328-48cd-9116-8fe63b268f14'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem 'c8623707-2d1d-4bbb-85cd-e45723cc6bfb'
ctrlr.c: 505:spdk_nvmf_ctrlr_connect: *ERROR*: Could not find subsystem '65736149-e3c0-4aa1-92e8-db62c80632e8'
ctrlr.c:1786:spdk_nvmf_ctrlr_process_admin_cmd: *ERROR*: Admin command sent before CONNECT
ctrlr.c:1786:spdk_nvmf_ctrlr_process_admin_cmd: *ERROR*: Admin command sent before CONNECT
[...]
```

**Comment by Roberciiik** [comment]
Also when I tried to disconnect a volume from initiator host (i.e. `nvme disconnect -d nvme4`) command line hangs and `dmesg` showed:
```
[859358.647049] INFO: task kworker/39:2:105471 blocked for more than 120 seconds.
[859358.661468]       Not tainted 4.10.0-28-generic #32~16.04.2-Ubuntu
[859358.673979] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[859358.689785] kworker/39:2    D    0 105471      2 0x00000000
[859358.689791] Workqueue: nvme_rdma_wq nvme_rdma_del_ctrl_work [nvme_rdma]
[859358.689792] Call Trace:
[859358.689797]  __schedule+0x232/0x700
[859358.689798]  schedule+0x36/0x80
[859358.689818]  blk_mq_freeze_queue_wait+0x4b/0xb0
[859358.689820]  ? wake_atomic_t_function+0x60/0x60
[859358.689821]  blk_mq_freeze_queue+0x1a/0x20
[859358.689822]  blk_freeze_queue+0xe/0x10
[859358.689842]  blk_cleanup_queue+0x7f/0x150
[859358.689846]  nvme_rdma_destroy_admin_queue+0x59/0xc0 [nvme_rdma]
[859358.689847]  nvme_rdma_shutdown_ctrl+0xc7/0xd0 [nvme_rdma]
[859358.689848]  __nvme_rdma_remove_ctrl+0x84/0x90 [nvme_rdma]
[859358.689849]  nvme_rdma_del_ctrl_work+0x1a/0x20 [nvme_rdma]
[859358.689852]  process_one_work+0x16b/0x4a0
[859358.689853]  worker_thread+0x4b/0x500
[859358.689855]  kthread+0x109/0x140
[859358.689856]  ? process_one_work+0x4a0/0x4a0
[859358.689861]  ? kthread_create_on_node+0x60/0x60
[859358.689865]  ret_from_fork+0x2c/0x40
[859358.689879] INFO: task nvme:108635 blocked for more than 120 seconds.
[859358.702904]       Not tainted 4.10.0-28-generic #32~16.04.2-Ubuntu
[859358.715436] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[859358.731244] nvme            D    0 108635 101463 0x00000004
[859358.731245] Call Trace:
[859358.731247]  __schedule+0x232/0x700
[859358.731247]  ? set_next_entity+0xc3/0x1b0
[859358.731248]  schedule+0x36/0x80
[859358.731249]  schedule_timeout+0x235/0x3f0
[859358.731250]  ? __schedule+0x23a/0x700
[859358.731251]  wait_for_completion+0xb4/0x140
[859358.731253]  ? wake_up_q+0x70/0x70
[859358.731254]  flush_work+0x129/0x1f0
[859358.731255]  ? worker_detach_from_pool+0xb0/0xb0
[859358.731256]  nvme_rdma_del_ctrl+0x60/0x90 [nvme_rdma]
[859358.731259]  nvme_sysfs_delete+0x2a/0x40 [nvme_core]
[859358.731278]  dev_attr_store+0x18/0x30
[859358.731281]  sysfs_kf_write+0x37/0x40
[859358.731282]  kernfs_fop_write+0x11d/0x1a0
[859358.731283]  __vfs_write+0x18/0x40
[859358.731283]  vfs_write+0xb8/0x1b0
[859358.731285]  ? do_sys_open+0x1b4/0x280
[859358.731286]  SyS_write+0x55/0xc0
[859358.731287]  entry_SYSCALL_64_fastpath+0x1e/0xad
[859358.731288] RIP: 0033:0x7faf31cda290
[859358.731289] RSP: 002b:00007ffc21bada88 EFLAGS: 00000246 ORIG_RAX: 0000000000000001
[859358.731291] RAX: ffffffffffffffda RBX: 000000000000002f RCX: 00007faf31cda290
[859358.731298] RDX: 0000000000000001 RSI: 0000000000427254 RDI: 0000000000000003
[859358.731301] RBP: 0000000000635280 R08: 000000000221dfe0 R09: 642f3131656d7600
[859358.731304] R10: 000000000000086f R11: 0000000000000246 R12: 0000000000000003
[859358.731307] R13: 00007ffc21baf7e2 R14: 00000000006354c0 R15: 00007ffc21bae060
```

**Comment by Seth5141** [comment]
@Roberciiik 
This morning I did a quick duplication on my dev system using fedorra 28 and a later linux kernel, and was unable to replicate the issue. That leads me to think that it has something to do with the kernel version/nvme cli you are using. Also, I have uploaded an RFC of the test script I am using so that you can verify that it is performing the same steps you are. (I am performing the steps in loopback though).
I then tried to recreate your system setup, but have been unable to get RDMA working on ubuntu 16.04 (especially with kernel version 4.10). I have installed the appropriate drivers and the MLNX_OFED iso, and have a connectx-4 NIC installed. Do you have any other configuration instructions or ideas for getting RDMA working on that OS version?
Also, just out of curiosity, Is there a specific reason you are working off of kernel version 4.10? The updated ubuntu 16.04.5 is currently on kernel 4.15.0.

Thanks,

Seth

**Comment by Roberciiik** [comment]
Hi,
I will try to upgrade my kernel. But it will take a few days probably (with testing)...

Btw,
It is my instruction for installation of OFED:

```
Mellanox OFED installation 4.4-1.0.0.0 (old version should be used 4.5)
•	download MLNX_OFED_LINUX-4.4-1.0.0.0-ubuntu16.04-x86_64.tgz from http://www.mellanox.com/page/products_dyn?product_family=26&mtag=linux_sw_drivers 
•	tar xvzf MLNX_OFED_LINUX-4.4-1.0.0.0-ubuntu16.04-x86_64.tgz 
•	cd MLNX_OFED_LINUX-4.4-1.0.0.0-ubuntu16.04-x86_64.tgz 
•	./mlnxofedinstall --add-kernel-support --with-nvmf --force 
•	apt install dracut-core 
•	dracut --force 
•	reboot
```

If it fail again, I will also provide a commands to check correctness of MLX installation.

**Comment by Seth5141** [comment]
@Roberciiik Thank you for the updated information. I am using a different version of the OFED than you were. Let me try again with this version following your commands to see if it works.

**Comment by Seth5141** [comment]
@Roberciiik Thanks again for the additional information. I now have a working system with the following characteristics:
ubuntu 16.04 - kernel version 4.10.0-28-generic
nvme-cli version 1.5
spdk 19.01.
I attempted to run the test script introduced in this patch https://review.gerrithub.io/c/spdk/spdk/+/446307. but it completed successfully without showing the errors you are seeing. 
I am going to copy the OS of the machine and try to use a couple of systems I have to do a full replication o fyour environment.
Can you please look at the test script I am using and let me know if it is properly duplicating the conditions you are seeing on your machine?

**Comment by Roberciiik** [comment]
@Seth5141 
Your steps are correct. But you have some differences which I would like to point out:
- I' using real NVMe drives (nvme_bdev)
- I'm try to connect from the other host machine instead of the same storage service host.
- In my test there was no destroy lvol bdev after disconnect but I think it is ok.

I have noticed that on the Storage Service machine I have kernel version 4.8.7.
Version 4.10 is on the initiator host.

**Comment by Roberciiik** [comment]
Hi, I reproduced the problem with newer kernel (4.15.0-45-generic).
I observed that after attachment of 3rd drive to initiator host, the first connected drive had gone.

**Comment by Seth5141** [comment]
thanks @Roberciiik I will be looking at your systems this afternoon or tomorrow morning.

**Comment by Seth5141** [comment]
@Roberciiik After looking at your setup this morning, I am pretty sure I have found the issue here. NVMe list does send some admin commands to the target side. These are then processed, and returned. There is however a known issue in spdk 19.01 with keeping track of the number of outstanding work requests we have in the RDMA qpair. What essentially happens is you perform a discovery or identify command that uses a buffer, then the keep_alive command reuses the same request object and the value is not cleared between uses. This results in the qpair incrementing a counter until it is impossible for future requests to get through. This results in a disconnect. 
I was able to verify this through the logs of the target (tail -f /tmp/nvmf_tgt.log | grep rdma.c makes the issue much clearer).
It is also shown in the intermittent nature of the failure. NVMe list tends to fail once, then work for some time before failing again.

The good news is that this fix has been backported to spdk-v19.01.1 which is about to be tagged. The current state of that point release can be checked out via the branch v19.01.x and it already contains my fix for this issue. I recommend checking out that branch and running your automated tests against it to verify that it completely solves your issue.

Thanks,

Seth

**Comment by Roberciiik** [comment]
Hi Seth,

Good news!
I cannot reproduce problem on the newer SPDK version.
I'll close the issue tomorrow if everything is OK.

Thanks

**Comment by Roberciiik** [comment]
Do you know when the official 19.01.1 tag will be assigned?

**Comment by Maciej-Szwed** [comment]
Hi Robert,
release will be this week.

**Comment by Seth5141** [comment]
@Roberciiik Great! Thanks for the feedback.

**Comment by Seth5141** [comment]
@robertciiik I am assuming everything went well? Closing this for now. Please feel free to reopen it if you find the same error, or open a new one for any different issues.
